### PR TITLE
Common interface for pair objectType and objectId

### PR DESCRIPTION
**Pull request by prodex**
see https://github.com/yiisoft-contrib/yiiframework.com/pull/236#discussion_r146107440

The interface is used in wiki, news, extension, comment, rating, star, file and contentShare.

Importing data from the old site (`ImportController`) was not able to test. Did not find where to get the DB of the old site. I think that my code does not destroy the import)

### CODE REVIEW

File: components/objectKey/ObjectKeyHelper.php [hunk 1]
```diff
@@ -0,0 +1,44 @@
...
+namespace app\components\objectKey;
+
+use app\models\Comment;
+use app\models\Extension;
+use app\models\File;
+use app\models\News;
+use app\models\Wiki;
+use yii\base\InvalidValueException;
+
+class ObjectKeyHelper
```
Comment (samdark) [inline components/objectKey/ObjectKeyHelper.php]: Could be named `ClassType`.

File: components/objectKey/ObjectKeyHelper.php [hunk 2]
```diff
@@ -0,0 +1,44 @@
...
+{
+    const TYPE_NEWS = 'news';
```
Comment (samdark) [inline components/objectKey/ObjectKeyHelper.php]: If class will be renamed to `ClassType`, these constants could be renamed to just `NEWS`, `WIKI` etc.

File: components/objectKey/ObjectKeyHelper.php [hunk 3]
```diff
@@ -0,0 +1,44 @@
...
+    const TYPE_WIKI = 'wiki';
+    const TYPE_EXTENSION = 'extension';
+    const TYPE_COMMENT = 'comment';
+    const TYPE_FILE = 'file';
+
+    const TYPE_GUIDE = 'guide';
+    const TYPE_API = 'api';
+
+    public static $typeClasses = [
```
Comment (samdark) [inline components/objectKey/ObjectKeyHelper.php]: Could be renamed to `$types`.

File: commands/ImportController.php [hunk 4]
```diff
@@ -33,6 +33,19 @@
  */
 class ImportController extends Controller
 {
+    /**
+     * @var string[]
+     */
+    public static $objectTypesMap = [
```
Comment (samdark) [inline commands/ImportController.php]: This map could be moved to `ObjectKeyHelper` itself.

File: components/objectKey/ObjectKeyHelper.php (hunk 1 above)
Comment (prodex) [reply components/objectKey/ObjectKeyHelper.php]: ok

File: components/objectKey/ObjectKeyInterface.php [hunk 5]
```diff
@@ -0,0 +1,16 @@
+<?php
+
+namespace app\components\objectKey;
+
+interface ObjectKeyInterface
```
Comment (samdark) [inline components/objectKey/ObjectKeyInterface.php]: `ObjectIdentityInterface`?

File: controllers/AjaxController.php [hunk 6]
```diff
@@ -108,10 +109,10 @@ public function actionVote($type, $id, $vote)
...
     {
         $model = null;
-        if (in_array($type, Star::$modelClasses, true)) {
-            /** @var $modelClass ActiveRecord */
-            $modelClass = "app\\models\\$type";
-            $model = $modelClass::findOne($id);
+        if (in_array($type, Star::$availableObjectTypes, true)) {
+            /** @var ActiveRecord $modelClass */
+            $modelClass = ObjectKeyHelper::getClass($type);
+            $model = $modelClass::findOne((int) $id);
```
Comment (samdark) [inline controllers/AjaxController.php]: No need to cast to int here.

File: commands/ImportController.php (hunk 4 above)
Comment (prodex) [reply commands/ImportController.php]: No, map is needed exclusively for import. I think a helper should not know about this.

File: views/wiki/view.php [hunk 7]
```diff
@@ -106,8 +107,8 @@
 <div class="comments-wrapper">
     <div class="container comments">
         <?= \app\widgets\Comments::widget([
-            'objectType' => \app\models\Wiki::COMMENT_TYPE,
-            'objectId' => "$model->id",
+            'objectType' => $model->getObjectType(),
+            'objectId' => (string) $model->getObjectId(),
```
Comment (samdark) [inline views/wiki/view.php]: No need to cast to string here.

File: components/objectKey/ObjectKeyInterface.php (hunk 5 above)
Comment (prodex) [reply components/objectKey/ObjectKeyInterface.php]: sounds better than ObjectKey)

File: components/objectKey/ObjectKeyHelper.php (hunk 3 above)
Comment (prodex) [reply components/objectKey/ObjectKeyHelper.php]: A variable returns classes. Maybe `$classes`?

File: views/wiki/view.php (hunk 7 above)
Comment (prodex) [reply views/wiki/view.php]: Fixed.

File: components/objectKey/ObjectKeyInterface.php (hunk 5 above)
Comment (prodex) [reply components/objectKey/ObjectKeyInterface.php]: Done.

File: controllers/AjaxController.php (hunk 6 above)
Comment (prodex) [reply controllers/AjaxController.php]: Done.

File: components/objectKey/ObjectKeyHelper.php (hunk 2 above)
Comment (prodex) [reply components/objectKey/ObjectKeyHelper.php]: Done.

File: commands/ImportController.php [hunk 8]
```diff
@@ -575,12 +592,16 @@ private function importComments()
 		$i = 0;
 		$err = 0;
 		foreach($query->each(100, $this->sourceDb) as $comment) {
-
 			try {
+			    $objectType = $this->convertObjectType($comment['object_type']);
+                if ($objectType === false) {
+                    throw new \Exception('Object type is "' . $comment['object_type'] . '" not found.');
```
Comment (samdark) [inline commands/ImportController.php]:
```php
'Object type "' . $comment['object_type'] . '" was not found.'
```

File: commands/ImportController.php [hunk 9]
```diff
@@ -680,8 +711,13 @@ public function importRatings()
 		$i = 0;
 		$err = 0;
 		foreach($ratingQuery->each(100, $this->sourceDb) as $rating) {
-
 			try {
+                $objectType = $this->convertObjectType($rating['object_type']);
+                if ($objectType === false) {
+                    throw new \Exception('Object type is "' . $rating['object_type'] . '" not found.');
```
Comment (samdark) [inline commands/ImportController.php]:
```php
'Object type "' . $rating['object_type'] . '" was not found.'
```

File: commands/ImportController.php [hunk 10]
```diff
@@ -742,8 +778,13 @@ public function importStars()
 		$i = 0;
 		$err = 0;
 		foreach($starQuery->each(100, $this->sourceDb) as $star) {
-
 			try {
+                $objectType = $this->convertObjectType($star['object_type']);
+                if ($objectType === false) {
+                    throw new \Exception('Object type is "' . $star['object_type'] . '" not found.');
```
Comment (samdark) [inline commands/ImportController.php]:
```php
'Object type "' . $star['object_type'] . '" was not found.'
```

File: components/object/ClassType.php [hunk 11]
```diff
@@ -0,0 +1,44 @@
...
+     *
+     * @return string
+     */
+    public static function getClass($type)
+    {
+        if (array_key_exists($type, static::$classes)) {
+            return static::$classes[$type];
+        }
+
+        throw new InvalidValueException("Type is '{$type}' not found.");
```
Comment (samdark) [inline components/object/ClassType.php]:
```php
'Object type "' . $star['object_type'] . '" was not found.'
```

File: notifications/views/comment-new.html.php [hunk 12]
```diff
@@ -13,7 +16,7 @@
 	'user' => $user,
 	'title' => "A new comment was added: " . $model->getLinkTitle(),
 ]) ?>
-	<p>A new comment was added to the following <?= $model->getItemType() ?> you are following:</p>
+	<p>A new comment was added to the following <?= $model->getObjectType() ?> you are following:</p>
```
Comment (samdark) [inline notifications/views/comment-new.html.php]:
```php
<p>A new comment was added to <?= Html::encode($model->getObjectType()) ?> you are following:</p>
```

File: notifications/views/comment-new.text.php [hunk 13]
```diff
@@ -13,7 +16,7 @@
 ]); ?>
 A new comment was added to the following content you are following:
```
Comment (samdark) [inline notifications/views/comment-new.text.php]: A new comment was added to the content you are following:

**Comment by prodex** [comment]
Done.

**Comment by samdark** [comment]
Thank you!

**Pull request merged**